File: front-end/src/app/domain/models/game/card/Card.ts
export enum CardType {
  CHEST = "CHEST",
  CHANCE = "CHANCE"
}

export abstract class Card {
  private _classType: string;
  private _id: number;
  private _description: string;
  private _type: CardType;

  protected constructor(
    classType: string,
    id: number,
    description: string,
    type: CardType
  ) {
    this._classType = classType;
    this._id = id;
    this._description = description;
    this._type = type;
  }

  public toJson(): object {
    return {
      "classType": this.classType,
      "id": this.id,
      "description": this.description,
      "type": this.type
    }
  }

  public get classType(): string {
    return this._classType;
  }

  public get id(): number {
    return this._id;
  }

  get description(): string {
    return this._description;
  }

  public get type(): CardType {
    return this._type;
  }

  public abstract toString(): string;
}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/card/Card.java
package nl.hva.backend.domain.models.game.card;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.GameBoard;
import nl.hva.backend.domain.models.game.board.location.Location;
import nl.hva.backend.domain.models.game.card.impl.*;

import javax.persistence.*;

/**
 * TODO JavaDoc
 *
 * @author Hamza el Haouti
 */
// Lombok
@Getter
@Setter
@NoArgsConstructor

// JPA
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)

// Jackson ORM config for inheritance.
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "classType"
)
@JsonSubTypes({
        @Type(value = FreeMoneyCard.class, name = "FreeMoneyCard"),
        @Type(value = GoToJailCard.class, name = "GoToJailCard"),
        @Type(value = JailBreakCard.class, name = "JailBreakCard"),
        @Type(value = MoveTokenCard.class, name = "MoveTokenCard"),
        @Type(value = MoveTokenToPositionCard.class, name = "MoveTokenToPositionCard"),
        @Type(value = PayMoneyCard.class, name = "PayMoneyCard"),
})
@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator.class,
        property = "id",
        scope = Card.class
)
public abstract class Card {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    private String description;

    @Enumerated(EnumType.STRING)
    private CardType type;

    @ManyToOne
    private GameBoard gameBoard;

    /**
     * TODO JavaDoc
     */
    public Card(
            String description,
            CardType type
    ) {
        this.description = description;
        this.type = type;
    }

    public enum CardType {
        CHEST,
        CHANCE
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Location location)) return false;
        return getId() == location.getId();
    }

    @Override
    public int hashCode() {
        return (int) (getId() ^ (getId() >>> 32));
    }
}